Move Add Contact Code to Core

stefaus opened the pull request:
This patch basically moves the existing add Contact logic into the core contact package.

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4586125&bid=15400347)

srossbach commented:
Wow this does reverts any effort we put in over years to remove UI related logic out of the core and out of the network layer.

stefaus commented:
First of all I don't think that handling xmpp error codes like in `getDialogContent(...)` is UI related logic, and I don't think the process of adding a contact is either.

What we hopefully can agree on is that IntelliJ should use the same logic in what steps happen to add a contact and how errors are handled, with the purpose of not reinventing this again and have a similar user experience.
As the same should be true for Saros Server I think the core is the only place for such code.

Please feel free to sketch a different solution.

srossbach commented on eclipse/src/saros/ui/wizards/AddContactWizard.java:
```diff
@@ -136,133 +98,24 @@ public JID getContact() {
-            + "Do you want to add the contact anyway?",
-        "Unknown error: " + e.getMessage());
+  private void addContact(JID jid, String nickname) throws SarosCancellationException {
+    BiPredicate<String, String> questionDialog =
```
Very questionable usage of a BiPredicate as a callback function, just because it works does not mean that it should be used that way.

srossbach commented on core/src/saros/net/xmpp/contact/AddContactHelper.java:
```diff
@@ -0,0 +1,148 @@
+    }
+  }
+
+  private static DialogContent getDialogContent(XMPPException e) {
```
Generating humanreadable error messages from error codes should be either part of the UI logic or some business logic in upper tiers and not in one of the lowest layer like the network. The whole code originally resided in the XMPPUtils class.

m273d15 replied:
I think this is the aspect we already talked about on other PRs: What do we want to do with IDE agnostic logic that provides information for the UI and is not pure core logic? This also contains the question: Which logic has to be the core?

Without blocking this PR we should answer these questions. I see the following approaches in order to separate the code:
1. separating the logic just in separate classes
1.  a separate package, the code is located in core
2. separating the core and core.ui in separate modules (req. JDK 9)
3.  a separate project with the ui code

I prefer approach 3 (or 2).
@stefaus  which approach do you prefer or do you have a totally different idea?

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/net/xmpp/contact/AddContactHelper.java  9
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/ui/wizards/AddContactWizard.java  -9
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/saros/saros/pullRequest?prid=4586125&bid=15400347)

stefaus commented on core/src/saros/net/xmpp/contact/AddContactHelper.java:
```diff
@@ -0,0 +1,148 @@
+import org.jivesoftware.smack.Connection;
+import org.jivesoftware.smack.XMPPException;
+import saros.SarosConstants;
+import saros.exceptions.SarosCancellationException;
```
OperationCanceledException is fine for me. How about moving the CancellationException into the negotiation package?

stefaus commented on core/src/saros/net/xmpp/contact/AddContactHelper.java:
```diff
@@ -0,0 +1,148 @@
+      BiPredicate<String, String> questionDialog)
+      throws SarosCancellationException {
+    Connection connection =
+        Objects.requireNonNull(connectionService.getConnection(), "connection is null");
```
I just moved that part with small adjustments, changing it is out of scope for now.

stefaus commented on core/src/saros/net/xmpp/contact/AddContactHelper.java:
```diff
@@ -0,0 +1,148 @@
+                + "Unfortunately your entered JID is unknown to the server.\n"
+                + "Please make sure you spelled the JID correctly.\n\n"
+                + "Do you want to add the contact anyway?")) {
+          throw new SarosCancellationException("Please make sure you spelled the JID correctly.");
```
I just moved that part with small adjustments, changing it is out of scope for now.

m273d15 commented on core/src/saros/net/xmpp/contact/AddContactHelper.java:
```diff
@@ -0,0 +1,148 @@
+import saros.net.xmpp.XMPPConnectionService;
+
+/**
+ * This class is mainly moved code from Eclipse {@code AddContactWizard} to add a contact on the
```
This information seems to be interesting in the context of the review, is this information still informative in the future. I think a class description is more useful.

m273d15 commented on core/src/saros/net/xmpp/contact/AddContactHelper.java:
```diff
@@ -0,0 +1,148 @@
+ * This class is mainly moved code from Eclipse {@code AddContactWizard} to add a contact on the
+ * XMPP Roster.
+ *
+ * <p>TODO This code has space for improvements, but probably update smack first..
```
If you already know the aspects you want to improve, please describe the aspects. Otherwise, it is not clear for other readers what you mean. You can also link an issue.

srossbach commented:
I tried to make an abstract

```
public interface IUIToolkit {

  public default <R> R runTask(ITask<R> task) throws InvocationTargetException {
    return runTask(task, false);
  }

  public <R> R runTask(ITask<R> task, boolean keepUIResponsive) throws InvocationTargetException;

  public <R> Future<R> runTaskAsync(ITask<R> task);

  public <R> R runTaskInContext(ITask<R> task, Object contextObject)
      throws InvocationTargetException;
}
```

However I would need @tobous help if IntelliJ a form of https://help.eclipse.org/neon/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fjface%2Foperation%2FIRunnableContext.html

It is still a bit questionable to work with context objects as it is up to the implementation how they are executed (in Eclipse for example using the WizardDialog will always mimic synchronous execution with the exception that if you do not fork the UI will become unresponsive, see https://help.eclipse.org/neon/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fjface%2Foperation%2FIRunnableContext.html)

tobous commented:
@stefaus I started to review this PR, but I am currently busy. Will have a look at it this evening or tomorrow.

@srossbach *Regarding the proposed interface:* I haven't had a more detailed look at the provided links yet, but in general, I am more than happy to collaborate on the design of the interface.

But I am not quite sure why the current design is on an abstraction level of running general UI tasks. 

My initial expectations/idea for the abstraction were more along the lines of dialog utils that offer specific kinds of dialogs (whichever are needed in the core). Implementations of the dialog util would then be provided by the IDE-specific implementations through dependency injection.
Although this might be more difficult as it initially seemed. Saros/I should be able to handle multiple project instances, which is why the [current dialog utils](https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/ui/util/SafeDialogUtils.java) get passed a project object. This is an IDE specific dependency that can't easily be resolved when dealing with a core interface.

*Regarding the `IRunnableContext`:* I just skimmed the article. Isn't that what we currently use the `IProgressMonitor` for?

> However I would need @tobous help if IntelliJ supports a form of https://help.eclipse.org/neon/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Freference%2Fapi%2Forg%2Feclipse%2Fjface%2Foperation%2FIRunnableContext.html

In general, I haven't dealt with that part of the logic enough to give you a straight answer right now. But from a short search, I found the [IntelliJ ProgressManager](https://github.com/JetBrains/intellij-community/blob/master/platform/core-api/src/com/intellij/openapi/progress/ProgressManager.java) that looks promising. It offers methods to run runables connected to a progress manager in a synchronous or asynchronous fashion.

srossbach commented:
Offering generic dialogs is not that hard. It can be done through an abstract factory. The Dialog interface mainly exists of one method that is basically 

```
int open();
```

As for the UI stuff. I already wrote a wrapper for a ProgressIndicator. Running UI tasks sync or async is not the challenging part. The problem is this Eclipse stuff where you can run such a tasks from a specific widget. The widget then provides the "running context" which is some thread logic and also supplies a specific monitor.

E.g when you run a runablecontext using a wizard the monitor is embedded in the wizard.

tobous commented:
> Offering generic dialogs is not that hard. It can be done through an abstract factory. The Dialog interface mainly exists of one method that is basically
> 
> ```
> int open();
> ```

I don't think it is that easy, at least with the current IntelliJ dialog logic. (Or I don't understand what the factory is supposed to build.)

Currently, the needed project object for a dialog is determined dynamically depending on the context. If a dialog is triggered through a UI element (like a button), the project the UI element is located in is used. If it is instead triggered by Saros, either the shared project is used if available (no idea what to do once multiple project can contain shared modules) or the most recently active project is used otherwise. A option would be to always use the most recently active project (at least for Saros-triggered dialogs), the big issue with this is that it can not be obtained synchronously from IntelliJ. (See [UIProjectUtils](https://github.com/saros-project/saros/blob/master/intellij/src/saros/intellij/ui/util/UIProjectUtils.java#L93).) As a result, the whole logic using the dialog would have to be passed run asynchronously as a runnable once the project object is available.

This issue would be an argument for the above suggested task abstraction for dialogs triggered externally (i.e. through UI buttons). But I am still not sure on that.

> The problem is this Eclipse stuff where you can run such a tasks from a specific widget. The widget then provides the "running context" which is some thread logic and also supplies a specific monitor.
> 
> E.g when you run a runablecontext using a wizard the monitor is embedded in the wizard.
> 
> The problem is that the Eclipse interface is so lazy when it comes to blocking. See the example above. If you specify fork it runs in the background but waits for termination of the tasks. When you use another implementation it may continue without waiting.

I am not 100% percent sure I understand the issue correctly. But doesn't this basically boil down to setting a standard in our interface (i.e. this method is blocking) and then making sure that the actual implementation meets this requirement?

tobous commented on core/src/saros/net/xmpp/contact/AddContactHelper.java:
```diff
@@ -0,0 +1,148 @@
+ *
+ * <p>TODO This code has space for improvements, but probably update smack first..
+ */
+class AddContactHelper {
```
Probably a matter of preference/taste, but I like `AddContactUtility` better. Just a suggestion though. If you prefer the current name, keeping it that way is fine with me as well.

```suggestion
class AddContactUtility {
```

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/net/xmpp/contact/AddContactUtility.java  9
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/ui/wizards/AddContactWizard.java  -9
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Saros/saros/pullRequest?prid=4586125&bid=15400347)

stefaus commented:
Incorporated some of the review points and added more infos for the TODO in the (renamed) Utility class. 

As we agreed to post pone the ui separation topic, I guess its ready for merging?

saros-infrastructure commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- core/src/saros/net/xmpp/contact/AddContactUtility.java  9
         

Complexity decreasing per file
==============================
+ eclipse/src/saros/ui/wizards/AddContactWizard.java  -9
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/Saros/saros/pullRequest?prid=4586125&bid=15400347)
